R/time.R

```diff
@@ -0,0 +1,197 @@
+#'   hours(1:5) + minutes(1:5) + seconds(1:5)
+#' )
+#' time_rec <- recipe(~ times, examples) %>%
+#'   step_time(all_predictors())
+#'
+#' tidy(time_rec, number = 1)
+#'
+#' time_rec <- prep(time_rec, training = examples)
+#'
+#' time_values <- bake(time_rec, new_data = examples)
+#' time_values
+#'
+#' tidy(time_rec, number = 1)
+step_time <-
+  function(recipe,
+           ...,
+           role = "predictor",
+           trained = FALSE,
+           features = c("hour24", "minute", "second"),
+           columns = NULL,
+           keep_original_cols = TRUE,
+           skip = FALSE,
+           id = rand_id("time")) {
+    feat <-
+      c(
+        "ampm",
+        "hour24",
+        "hour12",
+        "minute",
+        "second",
+        "decimal_day"
+      )
+    if (!is_tune(features) & !is_varying(features)) {
+      if (!all(features %in% feat)) {
+        rlang::abort(paste0(
+          "Possible values of `features` should include: ",
+          paste0("'", feat, "'", collapse = ", ")
+        ))
+      }
+    }
+    add_step(
+      recipe,
+      step_time_new(
+        terms = enquos(...),
+        role = role,
+        trained = trained,
+        features = features,
+        columns = columns,
+        keep_original_cols = keep_original_cols,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_time_new <-
+  function(terms, role, trained, features, columns, keep_original_cols, skip,
+           id) {
+    step(
+      subclass = "time",
+      terms = terms,
+      role = role,
+      trained = trained,
+      features = features,
+      columns = columns,
+      keep_original_cols = keep_original_cols,
+      skip = skip,
+      id = id
+    )
+  }
+
+
+#' @export
+prep.step_time <- function(x, training, info = NULL, ...) {
+  col_names <- recipes_eval_select(x$terms, training, info)
+
+  if (!all(purrr::map_lgl(training[, col_names], inherits, "POSIXt"))) {
+    rlang::abort(
+      paste0(
+        "All variables for `step_time` should be either `POSIXlt` or",
+        "`POSIXct` classes."
+      )
+    )
+  }
+
+  step_time_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    features = x$features,
+    columns = col_names,
+    keep_original_cols = get_keep_original_cols(x),
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+#' @export
+bake.step_time <- function(object, new_data, ...) {
+
+  for (column in object$columns) {
+    time_values <- get_time_features(
+      dt = getElement(new_data, column),
+      feats = object$features
+    )
+
+    names(time_values) <- glue::glue("{column}_{names(time_values)}")
+    new_data <- bind_cols(new_data, time_values)
+  }
+
+  keep_original_cols <- get_keep_original_cols(object)
+  if (!keep_original_cols) {
+    new_data <- new_data[, !(colnames(new_data) %in% object$columns), drop = FALSE]
+  }
+
+  new_data
+}
+
+get_time_features <- function(dt, feats) {
+
+  features <- list(
+    ampm = function(x) format(x, "%p"),
```

Comment:
Rather than returning a character vector here, I feel like it would be more appropriate to return a logical column that is `TRUE` if AM and `FALSE` if PM. Then you could just use `lubridate::am()` directly.

The logical column feels more useful for modeling? I think? Or maybe a binary integer column?

If you go this route, maybe it should be called `am` rather than `ampm`, or you could just document `ampm` well to ensure that people know which one corresponds to `TRUE`

Reply:
changing to am()